sherpa/models/basic.py

```diff
@@ -1783,125 +1785,291 @@ def calc(self, *args, **kwargs):
+        # this if the method has not changed but is it worth it?)
+        #
+        self.cache_clear()
+
     def __init__(self, name='tablemodel'):
         # these attributes should remain somewhat private
         # as not to conflict with user defined parameter names
+        #
+        # NOTE added: actually, we can not have user-defined names
+        # here thanks to the NoNewAttributesAfterInit base class.
+        #
         self.__x = None
         self.__y = None
         self.__filtered_y = None
         self.filename = None
-        self.method = linear_interp  # interpolation method
+        self._method = linear_interp
         self.ampl = Parameter(name, 'ampl', 1)
         ArithmeticModel.__init__(self, name, (self.ampl,))
 
-    def __setstate__(self, state):
-        self.__x = None
-        self.__y = state.pop('_y', None)
-        self.__filtered_y = state.pop('_filtered_y', None)
-        self.filename = state.pop('_file', None)
-        ArithmeticModel.__setstate__(self, state)
-
     def load(self, x, y):
-        self.__y = y
-        self.__x = x
+        """Set the model values.
+
+        Parameters
+        ----------
+        x, y : None or sequence
+           The model values. It is expected that either both are
+           given, and have the same number of elements, or that only y
+           is set, although the model can be cleared by setting both
+           to None.
+
+        """
+
+        if x is not None and y is None:
+            raise ModelErr("y must be set if x is set")
+
+        # Clear the cache. We could avoid doing this if the
+        # data has not changed but this is not worth the
+        # complexity.
+        #
+        self.cache_clear()
+
+        # A simplified version of sherpa.data._check, which is not
+        # used here to avoid circular dependencies. Rather than raise
+        # DataErr, we raise ModelErr with a similar message.
+        #
+        def _check(val):
+            if val is None:
+                return None
+
+            val = numpy.asarray(val)
+            if val.ndim != 1:
+                raise ModelErr("Array must be 1D or None")
 
-        # Input grid is sorted!
-        if x is not None:
-            idx = numpy.asarray(x).argsort()
-            self.__y = numpy.asarray(y)[idx]
-            self.__x = numpy.asarray(x)[idx]
+            return val
+
+        self.__y = _check(y)
+        self.__x = _check(x)
+
+        # clear the filtered array
+        self.__filtered_y = None
+
+        if x is None:
+            return
+
+        nx = len(self.__x)
+        ny = len(self.__y)
+        if nx != ny:
+            raise ModelErr(f"size mismatch between x and y: {nx} vs {ny}")
+
+        # Ensure the data is sorted. Is this useful?
+        #
+        idx = self.__x.argsort()
+        self.__y = self.__y[idx]
+        self.__x = self.__x[idx]
 
     def get_x(self):
+        """Return the independent axis, if set"""
```

Comment:
What does "if set" mean? It returns something (e.g. `None`) in any case, right? 
This is not a request to update the docstring, I just want to make sure I understand.

Reply:
"if set" here means "if not None, otherwise returns None". 

I always struggle with how much to say or not to say here, and I don't think I am consistent. In this particular case I don't think it's worth clarifying (as it is adding documentation to a routine which didn't have documentation ;-)

Reply:
For reference, "if set" means if `load` was called and `x` was not `None` in this call ;-)

Reply:
I've changed ", if set" to " or None" as I was editing the file anyway.